exercises/linked-list/src/example.c

```diff
@@ -1,148 +1,111 @@
-#include <stdlib.h>
-#include <stdbool.h>
 #include "linked_list.h"
+#include <assert.h>
+#include <stdlib.h>
 
-struct list_item {
-   struct list_item *next;
-   struct list_item *prev;
+struct list_node {
+   struct list_node *prev, *next;
    ll_data_t data;
 };
+struct list {
+   struct list_node *first, *last;
+};
 
-static struct list_item *new_item(ll_data_t data)
-{
-   struct list_item *item = malloc(sizeof(*item));
-
-   if (item != NULL) {
-      item->next = NULL;
-      item->prev = NULL;
-      item->data = data;
-   }
-   return item;
-}
-
-static struct list_item *get_list_tail(struct list_item **list)
-{
-   struct list_item *tail = NULL;
-   if (list != NULL) {
-      // climb list to tail
-      tail = *list;
-      while (tail->next != NULL) {
-         tail = tail->next;
-      }
-   }
-   return tail;
-}
-
-static void delete_item(struct list_item *item)
+static struct list_node *list_node_create(struct list_node *prev,
+                                          struct list_node *next,
+                                          ll_data_t data)
 {
-   if (item != NULL) {
-      free(item);
+   struct list_node *node = malloc(sizeof(*node));
+   if (node != NULL) {
+      node->prev = prev;
+      node->next = next;
+      node->data = data;
    }
+   return node;
 }
 
-struct list_item **new_list(void)
+struct list *list_create(void)
 {
-   struct list_item **list = malloc(sizeof(struct list_item *));
+   struct list *list = malloc(sizeof(*list));
    if (list != NULL) {
-      *list = NULL;
+      list->first = NULL;
+      list->last = NULL;
    }
    return list;
 }
 
-bool is_list_empty(struct list_item ** list)
+bool list_is_empty(const struct list *list)
 {
-   return (list == NULL || *list == NULL);
+   assert(list);
+   return list->first == NULL;
 }
 
-bool push(struct list_item ** list, ll_data_t item_data)
+struct list_node *list_push(struct list *list, ll_data_t data)
```

Comment:
This is a bit odd. Why does it return the pushed node?

Reply:
I agree, now that the layout of the `struct list` and `struct list_node` is hidden in the `.c` file it doesn't make sense anymore to return a pointer to the node.

Reply:
Thanks :+1: